src/main/kotlin/net/maizegenetics/phgv2/pathing/PathFinderWithViterbiHMM.kt

```diff
@@ -36,261 +37,206 @@ class PathFinderWithViterbiHMM(
+        val parent: PathNode?,
+        val sampleGametes: List<SampleGamete>,
+        val totalProbability: Double
+    )
+
+    init {
+        findLikelyParents = useLikelyParents && (maxParents < graph.sampleGametesInGraph().size || minCoverage < 1.0)
+        parentFinder = if (findLikelyParents) MostLikelyParents(graph) else null
+    }
 
-        init {
-            findLikelyParents = useLikelyParents && (maxParents < graph.sampleGametesInGraph().size || minCoverage < 1.0)
-            parentFinder = if (findLikelyParents) MostLikelyParents(graph) else null
-        }
+    fun findBestPath(readMap: Map<ReferenceRange, Map<List<String>, Int>>): Pair<List<PathNode>, List<MostLikelyParents.ParentStats>> {
+        //find likely parents, if required
+        val likelyParentList = if (findLikelyParents) parentFinder!!
+            .findMostLikelyParents(readMap, maxParents = maxParents, minCoverage = minCoverage)
+        else listOf()
+
+        val nodeList: MutableList<PathNode> = mutableListOf()
+        graph.contigs.forEach { chr ->
+            val finalNode = when {
+                isHaploidPath && findLikelyParents -> haploidViterbi(chr, readMap, likelyParentList.map { it.parent }.toSet())
+                isHaploidPath && !findLikelyParents -> haploidViterbi(chr, readMap)
+                !isHaploidPath && findLikelyParents -> diploidViterbi(chr, readMap, likelyParentList.map { it.parent }.toSet())
+                else -> diploidViterbi(chr, readMap)
+            }
 
-        /**
-         * For a map (list of hapids -> count of reads hitting the set), find the most likely path through the graph.
-         * If useLikelyParents = true, then a likely parents list is generated for each sample and only those parents are
-         * candiates for the path nodes for that sample.
-         *
-         * @param readMap a map of [ReferenceRange] to map(list of hapids to count of reads mapping to that list)
-         */
-        fun findBestHaploidPath(readMap: Map<ReferenceRange, Map<List<String>, Int>>): Pair<List<String>, List<MostLikelyParents.ParentStats>> {
-            val haplotypeList = mutableListOf<String>()
-            val likelyParentList = if (findLikelyParents) parentFinder!!
-                .findMostLikelyParents(readMap, maxParents = maxParents, minCoverage = minCoverage)
-            else listOf()
-
-            graph.contigs.forEach { chr ->
-                if (findLikelyParents) {
-                    haplotypeList.addAll(haploidViterbi(chr, readMap, likelyParentList.map { it.parent }.toSet()))
-                } else {
-                    haplotypeList.addAll(haploidViterbi(chr, readMap))
+            if (finalNode != null) {
+                val reverseNodeList = mutableListOf(finalNode)
+                var currentNode = finalNode!!
+                while (currentNode.parent != null) {
+                    currentNode = currentNode.parent!!
+                    reverseNodeList.add(currentNode)
```

Comment:
What is the use of the reverseNodeList?

Reply:
I am adding a comment to explain. Basically, because the nodes for a chromosome are added to the list in reverse order, last to first, I have to reverse the order before adding it to the complete list.